share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -0,0 +1,519 @@
+            );
+        }
+
+        // Send an empty response if the coordinator is not active
+        if (!isActive.get()) {
+            return CompletableFuture.completedFuture(
+                generateErrorWriteStateResponse(
+                    request,
+                    Errors.COORDINATOR_NOT_AVAILABLE,
+                    "Share coordinator is not available."
+                )
+            );
+        }
+
+        // The request received here could have multiple keys of structure group:topic:partition. However,
+        // the writeState method in ShareCoordinatorShard expects a single key in the request. Hence, we will
+        // be looping over the keys below and constructing new WriteShareGroupStateRequestData objects to pass
+        // onto the shard method.
+
+        request.topics().forEach(topicData -> {
+            Map<Integer, CompletableFuture<WriteShareGroupStateResponseData>> partitionFut =
+                futureMap.computeIfAbsent(topicData.topicId(), k -> new HashMap<>());
+            topicData.partitions().forEach(
+                partitionData -> partitionFut.put(partitionData.partition(), runtime.scheduleWriteOperation(
+                    "write-share-group-state",
+                    topicPartitionFor(SharePartitionKey.getInstance(groupId, topicData.topicId(), partitionData.partition())),
+                    Duration.ofMillis(config.shareCoordinatorOffsetsCommitTimeoutMs()),
+                    coordinator -> coordinator.writeState(context, new WriteShareGroupStateRequestData()
+                        .setGroupId(groupId)
+                        .setTopics(Collections.singletonList(new WriteShareGroupStateRequestData.WriteStateData()
+                            .setTopicId(topicData.topicId())
+                            .setPartitions(Collections.singletonList(new WriteShareGroupStateRequestData.PartitionData()
+                                .setPartition(partitionData.partition())
+                                .setStartOffset(partitionData.startOffset())
+                                .setLeaderEpoch(partitionData.leaderEpoch())
+                                .setStateEpoch(partitionData.stateEpoch())
+                                .setStateBatches(partitionData.stateBatches()))))))))
+            );
+        });
+
+        // Combine all futures into a single CompletableFuture<Void>
+        CompletableFuture<Void> combinedFuture = CompletableFuture.allOf(futureMap.values().stream()
+            .flatMap(partMap -> partMap.values().stream()).toArray(CompletableFuture[]::new));
+
+        return combinedFuture.thenApply(v -> {
+            List<WriteShareGroupStateResponseData.WriteStateResult> writeStateResults = futureMap.keySet().stream()
+                .map(topicId -> {
+                    List<WriteShareGroupStateResponseData.PartitionResult> partitionResults = futureMap.get(topicId).entrySet().stream()
+                        .map(topicEntry -> {
```

Comment:
The nested streams make the control flow here hard to follow. See if you can untangle it a bit. Maybe regular for loops would be more readable.